README.md
# pyramid_input
A Pyramid tween that normalizes HTTP request input data


CHANGELOG.rst
=========
ChangeLog
=========


v0.2.3
======

* Added support for re-parsing the request body for PATCH request
  methods (config ``pyramid_input.reparse-methods``)


v0.2.2
======

* Removed `distribute` dependency


v0.2.1
======

* Added logging of exceptions during key-value parse failures


v0.2.0
======

* First extricated and tagged release


pyramid_input/test.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# auth: Philip J Grabner <phil@canary.md>
# date: 2015/06/13
# copy: (C) Copyright 2015-EOT Canary Health, Inc., All Rights Reserved.
#------------------------------------------------------------------------------

import unittest
from collections import OrderedDict

from pyramid import testing
from pyramid.request import Request
from pyramid.httpexceptions import HTTPBadRequest
import six

from pyramid_input import factory, includeme

# todo: the ordering output of `repr` should be made explicit...
#       fix that so that it is dependable!

#------------------------------------------------------------------------------
class TestInputFactory(unittest.TestCase):

  def setUp(self):
    self.config = testing.setUp()
    self.registry = self.config.registry

  def test_register(self):
    handler = factory(None, self.registry)
    self.assertEqual(handler.__name__, 'input_tween')

#------------------------------------------------------------------------------
def canon(val):
  'Returns a canonical deep-copy version of `val`.'
  if isinstance(val, dict):
    return {canon(k): canon(val[k]) for k in sorted(val.keys(), key=repr)}
  if isinstance(val, (list, tuple)):
    return [canon(v) for v in val]
  if isinstance(val, six.string_types):
    try:
      return str(val)
    except UnicodeEncodeError:
      return val
  return val

#------------------------------------------------------------------------------
class TestInputTween(unittest.TestCase):

  def setUp(self):
    self.config = None
    self.registry = None
    self.tween = None
    self.request = None

  def setupTween(self, handler=None, settings=None):
    # includeme(self.config)
    self.config = testing.setUp()
    self.registry = self.config.registry
    self.registry.settings = settings or {}
    self.tween = factory(handler or self.handler, self.registry)

  def setupRequest(self, url='/', data=None, ctype=None, headers=None, params=None):
    if not self.tween:
      self.setupTween()
    if ctype:
      if not headers:
        headers = dict()
      headers['content-type'] = ctype
    request = Request.blank(url, headers=headers, POST=data, **(params or {}))
    self.request = request
    self.request.registry = self.registry

  def handler(self, request):
    return canon(request.input)

  def needYaml(self):
    try:
      import yaml
    except ImportError:
      raise unittest.SkipTest('yaml module not installed')

  #----------------------------------------------------------------------------

  def test_empty(self):
    self.setupRequest()
    self.assertEqual(self.tween(self.request), canon({}))

  def test_qs_only(self):
    self.setupRequest('/path?foo=bar')
    self.assertEqual(self.tween(self.request), canon({'foo': 'bar'}))

  def test_qs_unflatten(self):
    self.setupRequest('/path?foo.zig=bar&foo.zoo-0=e0&foo.zoo-1=e1')
    self.assertEqual(self.tween(self.request), canon({'foo': {'zig': 'bar', 'zoo': ['e0', 'e1']}}))

  def test_payload_only(self):
    self.setupRequest('/path', 'foo.zig=bar&foo.zoo-0=e0&foo.zoo-1=e1')
    self.assertEqual(self.tween(self.request), canon({'foo': {'zig': 'bar', 'zoo': ['e0', 'e1']}}))

  def test_qs_invalid(self):
    # note: unfortunately, formencode's `variable_decode` behaves differently
    #       in python 2 vs. python 3...
    # todo: do something about this! perhaps use `morph.unflatten()` instead?
    #       for now, a ticket was opened against formencode:
    #         https://github.com/formencode/formencode/issues/92
    self.setupRequest('/path', 'foo-1.zig=one&foo.0.zag=zog&foo-0.zig=two')
    if six.PY3:
      with self.assertLogs() as cm:
        self.assertEqual(
          self.tween(self.request),
          canon({'foo-1.zig': 'one', 'foo.0.zag': 'zog', 'foo-0.zig': 'two'}))
      self.assertEqual(len(cm.output), 1)
      self.assertEqual(
        cm.output[0].split('\n', 1)[0],
        'ERROR:pyramid_input:failed parsing key-value pairs')
    else:
      self.assertEqual(
        self.tween(self.request),
        canon({'foo': [{'zig': 'two'}, {'zig': 'one'}, {'zag': 'zog'}]}))

  def test_json(self):
    self.setupRequest('/path', '{"foo":{"zig":"bar","zoo":["e0","e1"]}}', ctype='application/json')
    self.assertEqual(self.tween(self.request), canon({'foo': {'zig': 'bar', 'zoo': ['e0', 'e1']}}))

  def test_json_invalid(self):
    self.setupRequest('/path', '{"foo":', ctype='application/json')
    ret = self.tween(self.request)
    self.assertEqual(ret.status_code, 400)
    self.assertEqual(str(ret), 'Invalid JSON')

  def test_yaml(self):
    self.needYaml()
    self.setupRequest('/path', 'foo:\n  zig: bar\n  zoo:\n  - e0\n  - e1', ctype='application/yaml')
    self.assertEqual(self.tween(self.request), canon({'foo': {'zig': 'bar', 'zoo': ['e0', 'e1']}}))

  def test_yaml_invalid(self):
    self.needYaml()
    self.setupRequest('/path', '{"foo":', ctype='application/yaml')
    ret = self.tween(self.request)
    self.assertEqual(ret.status_code, 400)
    self.assertEqual(str(ret), 'Invalid YAML')

  def test_xml(self):
    self.setupRequest('/path', '<foo zig="bar"><zoo>e0</zoo><zoo>e1</zoo></foo>', ctype='text/xml')
    self.assertEqual(self.tween(self.request), canon({'foo': {'zig': 'bar', 'zoo': ['e0', 'e1']}}))

  def test_xml_invalid(self):
    self.setupRequest('/path', '<foo>', ctype='application/xml')
    ret = self.tween(self.request)
    self.assertEqual(ret.status_code, 400)
    self.assertEqual(str(ret), 'Invalid XML')

  def test_xml_whitespace(self):
    self.setupRequest('/path', '''
<foo zig="bar">
  <zoo>e0</zoo>
  <zoo>e1</zoo>
  <space>  </space>
  <space idx='1'>
</space>
</foo>
''', ctype='text/xml')
    self.assertEqual(
      self.tween(self.request),
      canon({'foo': {'zig': 'bar', 'zoo': ['e0', 'e1'], 'space': ['  ', {'idx': '1', None: '\n'}]}}))

  def test_bad_contentType(self):
    self.setupRequest('/path', b'\x89PNG\r\n', ctype='image/png')
    ret = self.tween(self.request)
    self.assertEqual(ret.status_code, 400)
    self.assertEqual(str(ret), 'Unknown/unsupported content-type "image/png"')

  def test_combine_isolated(self):
    self.setupRequest('/path?foo=bar', 'zig=zag')
    self.assertEqual(self.tween(self.request), canon({'foo': 'bar', 'zig': 'zag'}))

  def test_combine_deep_leaves(self):
    self.setupRequest('/path?foo.a=b&foo.c=d', 'foo.c=f&foo.g=h')
    self.assertEqual(self.tween(self.request), canon({'foo': {'a': 'b', 'c': ['d', 'f'], 'g': 'h'}}))

  def test_combine_deep_branch(self):
    self.setupRequest('/path?foo.a=b&foo.c=d', 'foo=h')
    self.assertEqual(self.tween(self.request), canon({'foo': [{'a': 'b', 'c': 'd'}, 'h']}))

  def test_combine_conflict_merge(self):
    self.setupRequest('/path?foo=bar', 'foo=zag')
    self.assertEqual(self.tween(self.request), canon({'foo': ['bar', 'zag']}))

  def test_combine_conflict_override(self):
    self.setupTween(settings={'pyramid_input.combine.deep': False})
    self.setupRequest('/path?foo=bar', 'foo=zag')
    self.assertEqual(self.tween(self.request), canon({'foo': 'zag'}))

  def test_config_native_dict(self):
    def handler_aadict(request):
      return canon(request.input.zig.zag.zog)
    def handler_dict(request):
      return canon(request.input['zig']['zag']['zog'])
    self.setupTween(handler=handler_dict)
    self.setupRequest('/path?zig.zag.zog=bar')
    self.assertEqual(self.tween(self.request), canon('bar'))
    self.setupTween(handler=handler_aadict)
    self.setupRequest('/path?zig.zag.zog=bar')
    self.assertEqual(self.tween(self.request), canon('bar'))
    self.setupTween(handler=handler_dict, settings={'pyramid_input.native-dict': True})
    self.setupRequest('/path?zig.zag.zog=bar')
    self.assertEqual(self.tween(self.request), canon('bar'))
    self.setupTween(handler=handler_aadict, settings={'pyramid_input.native-dict': True})
    self.setupRequest('/path?zig.zag.zog=bar')
    with self.assertRaises(AttributeError) as cm:
      self.tween(self.request)
    self.assertEqual(str(cm.exception), "'dict' object has no attribute 'zig'")

  def test_config_attribute_name(self):
    def handler_data(request):
      return canon(request.data)
    self.setupTween(handler=handler_data)
    self.setupRequest('/path?foo=bar')
    with self.assertRaises(AttributeError) as cm:
      self.tween(self.request)
    self.assertEqual(str(cm.exception), "'Request' object has no attribute 'data'")
    self.setupTween(handler=handler_data, settings={'pyramid_input.attribute-name': 'data'})
    self.setupRequest('/path?foo=bar')
    self.assertEqual(self.tween(self.request), canon({'foo': 'bar'}))

  def test_config_enabled(self):
    self.setupRequest('/path?foo=bar')
    self.assertEqual(self.tween(self.request), canon({'foo': 'bar'}))
    self.setupTween(settings={'pyramid_input.enabled': False})
    self.setupRequest('/path?foo=bar')
    with self.assertRaises(AttributeError) as cm:
      self.tween(self.request)
    self.assertEqual(str(cm.exception), "'Request' object has no attribute 'input'")

  def test_config_exclude(self):
    def handler(request):
      return canon(getattr(request, 'input', None))
    self.setupTween(handler=handler, settings={'pyramid_input.exclude': '/a /b/**'})
    self.setupRequest('/path?foo=bar')
    self.assertEqual(self.tween(self.request), canon({'foo': 'bar'}))
    self.setupRequest('/a?foo=bar')
    self.assertEqual(self.tween(self.request), canon(None))
    self.setupRequest('/a/path?foo=bar')
    self.assertEqual(self.tween(self.request), canon({'foo': 'bar'}))
    self.setupRequest('/b?foo=bar')
    self.assertEqual(self.tween(self.request), canon({'foo': 'bar'}))
    self.setupRequest('/b/path/zoo?foo=bar')
    self.assertEqual(self.tween(self.request), canon(None))

  def test_config_include(self):
    def handler(request):
      return canon(getattr(request, 'input', None))
    self.setupTween(handler=handler, settings={'pyramid_input.include': '/a /a/**'})
    self.setupRequest('/path?foo=bar')
    self.assertEqual(self.tween(self.request), canon(None))
    self.setupRequest('/a?foo=bar')
    self.assertEqual(self.tween(self.request), canon({'foo': 'bar'}))
    self.setupRequest('/a/path?foo=bar')
    self.assertEqual(self.tween(self.request), canon({'foo': 'bar'}))

  def test_config_require_dict(self):
    self.needYaml()
    self.setupRequest('/path', 'foo: bar', ctype='application/yaml')
    self.assertEqual(self.tween(self.request), canon({'foo': 'bar'}))
    self.setupRequest('/path', '["foo", "bar"]', ctype='application/yaml')
    ret = self.tween(self.request)
    self.assertEqual(ret.status_code, 400)
    self.assertEqual(str(ret), 'Payload must evaluate to dict-type')
    self.setupTween(settings={'pyramid_input.require-dict': False})
    self.setupRequest('/path', 'foo: bar', ctype='application/yaml')
    self.assertEqual(self.tween(self.request), canon({'foo': 'bar'}))
    self.setupRequest('/path', '["foo", "bar"]', ctype='application/yaml')
    self.assertEqual(self.tween(self.request), canon(['foo', 'bar']))

  def test_config_fail_unknown(self):
    self.setupRequest('/path', b'\x89PNG\r\n', ctype='image/png')
    ret = self.tween(self.request)
    self.assertEqual(ret.status_code, 400)
    self.assertEqual(str(ret), 'Unknown/unsupported content-type "image/png"')
    self.setupTween(settings={'pyramid_input.fail-unknown': False})
    self.setupRequest('/path', b'\x89PNG\r\n', ctype='image/png')
    self.assertEqual(self.tween(self.request), canon({}))

  def test_config_error_handler(self):
    def error_handler(request, error):
      import json
      request.response.status_code = error.code
      request.response.content_type = 'application/json'
      request.response.body = json.dumps(OrderedDict([
        ('status', error.code),
        ('message', str(error)),
      ])).encode('ascii')
      return request.response
    self.setupTween(settings={'pyramid_input.error-handler': error_handler})
    self.setupRequest('/path', '<|>-+-<|>', ctype='application/xml')
    ret = self.tween(self.request)
    self.assertEqual(ret.status_code, 400)
    self.assertEqual(ret.content_type, 'application/json')
    self.assertEqual(ret.body, b'{"status": 400, "message": "Invalid XML"}')

  def test_method_patch(self):
    self.setupRequest(
      '/path',
      ctype   = 'application/x-www-form-urlencoded',
      params  = dict(
        method  = 'PATCH',
        body    = 'foo.zig=bar&foo.zoo-0=e0&foo.zoo-1=e1'
      ))
    self.assertEqual(self.tween(self.request), canon({'foo': {'zig': 'bar', 'zoo': ['e0', 'e1']}}))


#------------------------------------------------------------------------------
# end of $Id$
# $ChangeLog$
#------------------------------------------------------------------------------


pyramid_input/__init__.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# auth: Philip J Grabner <phil@canary.md>
# date: 2015/06/13
# copy: (C) Copyright 2015-EOT Canary Health, Inc., All Rights Reserved.
#------------------------------------------------------------------------------

import logging

from pyramid.settings import aslist, asbool
from pyramid.httpexceptions import HTTPException, HTTPBadRequest
import asset
from aadict import aadict
import globre
import formencode
import morph
from six.moves.urllib.parse import parse_qs
try:
  import yaml
except ImportError:
  yaml = None

from .i18n import _

#------------------------------------------------------------------------------

log = logging.getLogger(__name__)

CONFIG_PREFIX               = 'pyramid_input.'
DEFAULT_ATTRIBUTE_NAME      = 'input'
DEFAULT_REPARSE_METHODS     = 'PATCH',

FORM_CTYPES_PREFIX          = 'application/x-www-form-urlencoded', 'multipart/form-data',
JSON_CTYPES_PREFIX          = 'application/json', 'application/x-json', 'text/json', 'text/x-json'
JSON_CTYPES_SUFFIX          = '+json',
YAML_CTYPES_PREFIX          = 'application/yaml', 'application/x-yaml', 'text/yaml', 'text/x-yaml'
YAML_CTYPES_SUFFIX          = '+yaml',
XML_CTYPES_PREFIX           = 'application/xml',  'application/x-xml',  'text/xml',  'text/x-xml'
XML_CTYPES_SUFFIX           = '+xml',

ERRMSG_UNSUPPORTED          = 'Unknown/unsupported content-type "{content_type}"'
ERRMSG_DICTTYPE             = 'Payload must evaluate to dict-type'
ERRMSG_INVJSON              = 'Invalid JSON'
ERRMSG_INVYAML              = 'Invalid YAML'
ERRMSG_INVXML               = 'Invalid XML'

#------------------------------------------------------------------------------
def factory(handler, registry):
  get          = morph.pick(registry.settings, prefix=CONFIG_PREFIX).get
  conf         = aadict()
  conf.enabled = asbool(get('enabled', True))
  conf.include = [globre.compile(el, globre.EXACT)
                  for el in aslist(get('include', []))]
  conf.exclude = [globre.compile(el, globre.EXACT)
                  for el in aslist(get('exclude', []))]
  conf.reparse = aslist(get('reparse-methods', DEFAULT_REPARSE_METHODS))
  conf.name    = get('attribute-name', DEFAULT_ATTRIBUTE_NAME)
  conf.deep    = asbool(get('combine.deep', True))
  conf.reqdict = asbool(get('require-dict', True))
  conf.failunk = asbool(get('fail-unknown', True))
  conf.ndict   = asbool(get('native-dict', False))
  conf.error   = get('error-handler', None)
  if conf.error:
    conf.error = asset.symbol(conf.error)
  conf.xfmt    = asbool(get('xml.enable', True))
  conf.jfmt    = asbool(get('json.enable', True))
  conf.yfmt    = asbool(get('yaml.enable', bool(yaml or get('yaml.parser'))))
  if conf.jfmt:
    conf.jparser = get('json.parser', None)
    if conf.jparser:
      conf.jparser = asset.symbol(conf.jparser)
  if conf.yfmt:
    conf.yparser = asset.symbol(get('yaml.parser', 'yaml.load'))
  if conf.xfmt:
    conf.xparser = asset.symbol(get('xml.parser', 'xml.etree.ElementTree.fromstring'))
  def input_tween(request):
    return process(handler, request, conf)
  return input_tween

#------------------------------------------------------------------------------
def process(handler, request, conf):
  try:
    decorate(request, conf)
  except HTTPException as exc:
    if conf.error:
      return conf.error(request, exc)
    return exc
  return handler(request)

#------------------------------------------------------------------------------
def decorate(request, conf):
  if not conf.enabled:
    return
  if conf.include:
    for el in conf.include:
      if el.match(request.path):
        break
    else:
      return
  if conf.exclude:
    for el in conf.exclude:
      if el.match(request.path):
        return
  setattr(request, conf.name, compute(request, conf))

#------------------------------------------------------------------------------
def isContentType(request, prefixes=None, suffixes=None):
  ctype = request.content_type.split(';', 1)[0].split(',', 1)[0]
  for ct in prefixes or []:
    if ctype.startswith(ct):
      return True
  for ct in suffixes or []:
    if ctype.endswith(ct):
      return True
  return False

#------------------------------------------------------------------------------
def mergeInto(base, items, deep=True):
  for k, v in items:
    if k in base:
      if deep and isinstance(v, dict) and isinstance(base[k], dict):
        base[k] = mergeInto(base[k], v.items(), True)
      else:
        if not isinstance(base[k], list):
          base[k] = [base[k]]
        base[k].append(v)
    else:
      base[k] = v
  return base

#------------------------------------------------------------------------------
def compute(request, conf):
  if conf.ndict:
    return _compute(request, conf)
  return aadict.d2ar(_compute(request, conf))

#------------------------------------------------------------------------------
def _compute(request, conf):
  qs = parse_pairs(request.GET, conf)
  pay = None
  if request.content_type and request.content_length:
    if isContentType(request, FORM_CTYPES_PREFIX):
      if conf.reparse and request.method in conf.reparse:
        pay = parse_form(request, conf)
      else:
        pay = parse_pairs(request.POST, conf)
    elif conf.jfmt and isContentType(request, JSON_CTYPES_PREFIX, JSON_CTYPES_SUFFIX):
      pay = parse_json(request, conf)
    elif conf.yfmt and isContentType(request, YAML_CTYPES_PREFIX, YAML_CTYPES_SUFFIX):
      pay = parse_yaml(request, conf)
    elif conf.xfmt and isContentType(request, XML_CTYPES_PREFIX, XML_CTYPES_SUFFIX):
      pay = parse_xml(request, conf)
    else:
      if conf.failunk:
        raise HTTPBadRequest(_(ERRMSG_UNSUPPORTED).format(
          content_type=request.content_type))
  if pay:
    if conf.reqdict and not morph.isdict(pay):
      raise HTTPBadRequest(ERRMSG_DICTTYPE)
    if not qs or not morph.isdict(pay):
      return pay
    if conf.deep:
      qs = mergeInto(qs, pay.items(), True)
    else:
      qs.update(pay)
  return qs

#------------------------------------------------------------------------------
def parse_form(request, conf):
  return parse_items([
    (key, val)
    for key, vals in parse_qs(request.body or '').items()
    for val in vals
  ], conf)

#------------------------------------------------------------------------------
def parse_pairs(pairs, conf):
  return parse_items(pairs.items(), conf)

#------------------------------------------------------------------------------
def parse_items(items, conf):
  ret = mergeInto(dict(), items, False)
  try:
    return dict(formencode.variabledecode.variable_decode(ret))
  except Exception as exc:
    log.exception('failed parsing key-value pairs')
    return ret

#------------------------------------------------------------------------------
def parse_json(request, conf):
  try:
    if conf.jparser:
      return conf.jparser(request.body)
    return request.json_body
  except ValueError:
    log.exception('failed parsing JSON')
    raise HTTPBadRequest(_(ERRMSG_INVJSON))

#------------------------------------------------------------------------------
def parse_yaml(request, conf):
  try:
    return conf.yparser(request.body)
  except Exception:
    log.exception('failed parsing YAML')
    raise HTTPBadRequest(_(ERRMSG_INVYAML))

#------------------------------------------------------------------------------
def node2pair(node):
  ret = dict(node.items())
  if node.text:
    if len(ret) <= 0 and len(node) <= 0:
      return (node.tag, node.text)
    if len(node) <= 0 or node.text.strip():
      ret[None] = node.text
  mergeInto(ret, [node2pair(n) for n in node])
  return (node.tag, ret)

#------------------------------------------------------------------------------
def parse_xml(request, conf):
  try:
    doc = conf.xparser(request.body)
  except Exception:
    log.exception('failed parsing XML')
    raise HTTPBadRequest(_(ERRMSG_INVXML))
  return dict([node2pair(doc)])

#------------------------------------------------------------------------------
def includeme(config):
  '''
  Adds a pyramid :term:`tween` to `config` that parses and combines
  all request data sources into a single data tree. See
  `https://github.com/canaryhealth/pyramid_input`_ for details.
  '''
  if asbool(config.registry.settings.get(CONFIG_PREFIX + 'enabled', True)):
    config.add_tween('pyramid_input.factory')

#------------------------------------------------------------------------------
# end of $Id$
# $ChangeLog$
#------------------------------------------------------------------------------
